core/packages/contracts/contracts/NativeTokens.sol

```diff
@@ -0,0 +1,123 @@
+    /// @param _vault The vault to use to `lock`/`unlock` tokens.
+    /// @param _outboundChannel The channel used to queue lock and create messages.
+    constructor(ERC20Vault _vault, OutboundChannel _outboundChannel) {
+        vault = _vault;
+        outboundChannel = _outboundChannel;
+    }
+
+    /// @notice Locks tokens to mint on substrate.
+    /// @dev Locks an amount of ERC20 Tokens in the vault and enqueues a mint message. Requires the allowance to be set on the ERC20 token where the spender is the vault.
+    /// @param token The token to lock.
+    /// @param recipient The recipient on the substrate side.
+    /// @param amount The amount to lock.
+    function lock(address token, bytes32 recipient, uint256 amount) public {
+        require(amount > 0, "NativeTokes: non zero amount");
+        require(token != address(0), "NativeTokes: non zero address");
+        require(recipient != 0, "NativeTokes: non zero recipient");
+
+        /// TODO: Implement a max locked amount.
+        vault.deposit(msg.sender, token, amount);
+
+        /// TODO: Encode a call
+        bytes memory call;
+        /// TODO: Get weight
+        uint64 weight = 1_000_000;
+
+        outboundChannel.submit(msg.sender, call, weight);
+        emit Locked(msg.sender, recipient, token, amount);
+    }
+
+    /// @notice Creates a native token.
+    /// @dev Enqueues a create native token message to substrate.
+    /// @param token The ERC20 token address.
+    /// @param name The name of the ERC20 token.
+    /// @param symbol The symbol of the ERC20 token.
+    /// @param decimals The decimals for the ERC20 token.
+    function create(
+        address token,
+        string calldata name,
+        string calldata symbol,
+        uint8 decimals
+    ) public {
+        /// TODO: Encode a call
+        bytes memory call;
+        /// TODO: Get weight
+        uint64 weight = 1_000_000;
+
+        outboundChannel.submit(msg.sender, call, weight);
+        emit Created(token);
+    }
+
+    /// @notice Handles messages coming in over the bridge.
+    /// @dev Processes messages from inbound channel.
+    /// @param origin The hashed substrate sovereign account.
+    /// @param message The message enqueued from substrate.
+    function handle(bytes32 origin, bytes calldata message) external onlyOwner {
+        /// TODO: require origin is statemint.
+        MessageProtocol.Message memory decoded = abi.decode(message, (MessageProtocol.Message));
+        if (decoded.action == MessageProtocol.Action.Unlock) {
+            unlock(origin, abi.decode(decoded.payload, (UnlockPayload)));
+        } else {
+            revert("NativeTokens: unknown action");
+        }
+    }
+
+    /// @notice Unlocks funds from the vault and sends it to recipient.
+    /// @param origin The hashed substrate sovereign account.
+    /// @param payload A decoded unlock payload.
+    function unlock(bytes32 origin, UnlockPayload memory payload) private {
```

Comment:
I'd prefer to call this `doUnlock`, which kinda implies that its an internal method for handling something.